stabilization: Introduce control input smoothing

glowtape opened the pull request:
@mlyle and I discussed this a few months ago (I did some work, however a lot of procrastrination happened). The idea is to take the stair stepped RC input signal and smooth the corners out a bit, by using some amount of prediction to prevent delay. The output signal of it would have the leading edges chamfered. This is supposed reduce the P- and D-term kicks on large control input changes, and in consequence reduce actuator saturation.

This early implementation has four modes. The chamfered steps with two different prediction duty cycles, simple linear interpolation (for whoever wants it, maybe planes for servo smoothness), and interpolation by using predicted values similar to the first two modes (same as previous, but with reduced delay).

In normal operation, it should maximally only predict as long as it detects the frame timing to be (which is per-axis, for the very rare multi-RX setup). There's a hard limit of 20ms, since that's the PPM timing, in case something goes wrong with detection of the timing (it stores the smallest number it encounters).

The final implementation would have modes selectable per axis for RPY+T at least via UAVO browser.

Right now only smoothcontrol.c/h in this PR are of interest, the rest is just barebones integration for Seppuku.

Here's what the generated signals look in the simulator code (the green stuff is FFT, forgot why I added that the last time).

Chamfered steps with 50% of the signal time predicted.
![icee50](https://cloud.githubusercontent.com/assets/4010813/24062402/b85c8200-0b5b-11e7-82d9-4d1f0c5433e6.png)

Same but with 75% duty cycle.
![icee75](https://cloud.githubusercontent.com/assets/4010813/24062425/cfb4f66c-0b5b-11e7-981c-c6d754bebd9d.png)

Linear interpolation, well, result as expected.
![interpol](https://cloud.githubusercontent.com/assets/4010813/24062434/d9b04d38-0b5b-11e7-8abf-7a171761fe83.png)

Interpolation using predicts. With current parameters, the overshoots are pretty tame.
![predictive](https://cloud.githubusercontent.com/assets/4010813/24062449/e2b446be-0b5b-11e7-8aed-862025c83b1e.png)

In my opinion, the 75% one seems to be a nice approach, with the interpolated version for least delay maximum smoothness. The overshoots in the former are very very tame compared to plain extrapolation. And they look unspectacular in latter, too. (If you switch between 75% and interpolated predicts, the curve looks similar enough.)

glowtape committed:
stabilization: Introduce control input smoothing.

glowtape commented on flight/Libraries/math/smoothcontrol.c:
```diff
@@ -0,0 +1,194 @@
+		case SMOOTHCONTROL_NORMAL:
+		case SMOOTHCONTROL_EXTENDED:
+			axis->differential = signal_diff * SMOOTHCONTROL_PREDICTOR_SLOPE / (float)axis->control_interval;
+			axis->current = new_signal + signal_diff * SMOOTHCONTROL_CHAMFER_START;
```
new_signal should be axis->signal. Has been fixed.

glowtape commented:
Fixed a bunch of stupid mistakes and did a living room test. Seems to do no harm at least.

I did have a few odd twitches during hover, but that might be clipping related (IIRC the last time that happened, I had to change a motor, need to diagnose with Openlager).

glowtape commented:
This OK? Code belongs to stab.c, but it makes more sense in the input configuration.

(Not sure why this empty label shows up near the buttons, I didn't touch anything outside input.ui.)

![smoothingui](https://cloud.githubusercontent.com/assets/4010813/24081125/f9637216-0cac-11e7-984b-aa35adbc5f2e.png)

mlyle commented on flight/Libraries/math/smoothcontrol.c:
```diff
@@ -0,0 +1,195 @@
+	// Stores the lowest amount of ticks between control input updates.
+	// uint8_t allows for 32ms at 8KHz, should be sufficient to deal with
+	// even PPM and PWM at that rates (plz no).
+	uint8_t control_interval;
```
I'm concerned jitter has the possibility to effectively disable this.  Maybe IIR this?

glowtape replied:
Jitter didn't appear much of an issue. If at all, the signal is rather a tad late than on time. The only issue was when the stab loop started, it immediately converged to 1. Thus the stuff on L125.

mlyle replied:
If things happen at 9ms intervals, and something happens that ties up the FC for 5 ms, we could get a 14ms interval and then a 4ms one.  Then afterwards our behavior would be permanently changed.

This is probably part of how you get the 1ms at first.

tracernz replied:
Also nothing wrong with PPM or PWM so `(plz no)` isn't really necessary.

mlyle commented on flight/Libraries/math/smoothcontrol.c:
```diff
@@ -0,0 +1,195 @@
+			// Doesn't do anything in the integrator.
+			break;
+
+		// icee's proposal, creates chamfered steps with signal prediction
```
We need a worst-case timeout that applies no matter what.  On the order of 15ms.

mlyle replied:
(Which is somewhat difficult here because it's in terms of control Hz)

glowtape replied:
There's a "time bomb" being calculated down in L179. It uses a #define that's currently set to 20ms (typical PPM timing) and derives an amount of ticks based on dT, then gets applied in L139. If the time bomb is passed, the original signal is forced. dT_expected is being passed from stab.c, so it should always be some valid value, and if actual dT doesn't match, it'll create a critical alarm and doesn't let the user arm.

mlyle commented on flight/Libraries/math/smoothcontrol.c:
```diff
@@ -0,0 +1,195 @@
+			break;
+
+		// Linear interpolation of predicts of the type above.
+		case SMOOTHCONTROL_SUPERSMOOTH:;
```
While it's not an issue now, some compilers don't like `label: ;`    while `(void) 0;`  is guaranteed to work, though wordier.

tracernz replied:
In the rest of our code (both flight and GCS) we use scopes `{ }`. I think we should stick with that rather than introducing yet another style.

mlyle replied:
Hm, there's a bunch of after-first-statement variable decls in cases too.  I dunno.

mlyle commented on flight/Libraries/math/smoothcontrol.c:
```diff
@@ -0,0 +1,195 @@
+	// handle it separately.
+	if(new_signal < 0)
+	{
+		float tmp = 0;
```
really really need to think about the safety of this, especially with fancy interpolators.  Can a negative throttle signal ever be interpolated to positive?

mlyle replied:
(I believe it can for awhile with the chamfering stuff)

glowtape replied:
Yea. When you pass -1 as-is into the code and it stores it as last value, and then you throttle up (throttle range is 0..1), it'll predict a large value, which causes the motors to accelerate hard for a control interval, just to brake like hell when there's proper prediction. I didn't check for -1 on an earlier version, so the quad always made a *clunk* noise. It doesn't write back to new_signal on -1 and tells the throttle axis state that there's 0 throttle.

mlyle replied:
Right.  But doesn't this mean that a negative throttle command can become a positive one (briefly, at least)?  E.g. throttle was 1, stick is moved to -1, which is treated as 0 here.  Then in turn we go "most of the way there" but don't actually disable throt...

glowtape replied:
new_signal might probably be an irritating variable name. A value gets fed into the function via new_signal, some math happens and it gets dumped back into new_signal (the raw_input and ActuatorDesired.Thrust fields, that are processed, get set up anew in stab.c every loop, anyway).

In this wrapper, when it crosses into negative, a zero gets forced into the predictor, while the value in new_signal will be left untouched. So when throttle changes from say 0.5 to -1 over a receiver update, it'll leave the value at -1 and predict into nirvana (the tmp float). The last known signal in the state gets set to zero by feeding tmp=0, and when throttle crosses back into positive, it'll start doing its predicting in an orderly fashion, starting from zero.

mlyle commented on flight/Modules/Stabilization/stabilization.c:
```diff
@@ -1021,6 +1043,23 @@ static void zero_pids(void)
+	switch(m)
+	{
+		default:
+		case STABILIZATIONSETTINGS_RCCONTROLSMOOTHING_NONE:			return SMOOTHCONTROL_NONE;
```
newlines/indentiness..

mlyle commented on flight/targets/pipxtreme/fw/Makefile:
```diff
@@ -175,6 +175,7 @@ SRC += $(FLIGHTLIB)/circqueue.c
 SRC += $(FLIGHTLIB)/morsel.c
 SRC += $(MATHLIB)/misc_math.c
 SRC += $(MATHLIB)/lpfilter.c
+SRC += $(MATHLIB)/smoothcontrol.c
```
Shouldn't be here, right?

mlyle reviewed (commented):
A very good start, though having lots of options makes me nervous.

glowtape commented:
Hmmm, I'm just out on the field and it freaks out. This is on hold them I guess.

glowtape commented:
The code behaves properly again.

glowtape commented:
Flew a bunch of packs on this configured to Supersmooth mode on all axes. Flies smoothly.

The twitches mentioned a few days ago are gone, they happened due to forgetting to dereference a pointer.

Rebased to merge cleanly with the stab.c changes.

glowtape commented:
I think this still needs bounding the predictions to the max. rates.

And probably a limit on how much deg/s it's allowed predict ahead of the current known signal, as to prevent huge overshoot and subsequent pull back that might happen on the end of large stick throws on a slow receiver. ESCs would probably thank for that.

I'd say 90deg/s might be a nice limit?

glowtape commented:
Flew all packs of mine on the sloped step version with 75% prediction duty cycle this time. No weird stuff either.

tracernz commented on flight/Libraries/math/smoothcontrol.c:
```diff
@@ -0,0 +1,217 @@
+#include "pios.h"
+#include "smoothcontrol.h"
+
+#define ASSERT(x) if(x) PIOS_Assert(0);
```
Why not just use PIOS_Assert? This macro actually inverts the meaning of ASSERT which is a very bad idea IMO. If you really must keep it, it should be called ASSERT_FALSE.

tracernz commented on flight/Libraries/math/smoothcontrol.c:
```diff
@@ -0,0 +1,217 @@
+#include "smoothcontrol.h"
+
+#define ASSERT(x) if(x) PIOS_Assert(0);
+#define MAX(v,x) ((v) > (x) ? (x) : (v))
```
We should probably have a good min/max implementation in pios.h or openpilot.h rather than random ones like this. e.g. http://stackoverflow.com/a/3437484. But I thought arm-none-eabi headers already give us MIN/MAX??

tracernz replied:
We do already have higher quality ones: https://github.com/d-ronin/dRonin/blob/ebb1954b074bc168491bf11bdb639ddced14a0f9/flight/Libraries/math/misc_math.h#L38

glowtape commented:
Rebased and condensed.

mlyle commented on flight/Libraries/math/smoothcontrol.c:
```diff
@@ -0,0 +1,178 @@
+			axis->differential = signal_diff * SMOOTHCONTROL_PREDICTOR_SLOPE / (float)state->control_interval;
+			axis->current = axis->signal + signal_diff * SMOOTHCONTROL_CHAMFER_START;
+			axis->integrator_timeout = (uint8_t)MAX((float)state->control_interval*
+				(axis->mode == SMOOTHCONTROL_NORMAL ? SMOOTHCONTROL_DUTY_CYCLE : SMOOTHCONTROL_EXTENDED_DUTY_CYCLE), 255);
```
If 255 is chosen, does this ever time out?

mlyle commented on flight/Libraries/math/smoothcontrol.c:
```diff
@@ -0,0 +1,178 @@
+		case SMOOTHCONTROL_EXTENDED:
+			axis->differential = signal_diff * SMOOTHCONTROL_PREDICTOR_SLOPE / (float)state->control_interval;
+			axis->current = axis->signal + signal_diff * SMOOTHCONTROL_CHAMFER_START;
+			axis->integrator_timeout = (uint8_t)MAX((float)state->control_interval*
```
space before star

mlyle commented on flight/Libraries/math/smoothcontrol.c:
```diff
@@ -0,0 +1,178 @@
+	PIOS_Assert(state);
+
+	if(!*state) {
+		*state = PIOS_malloc_no_dma(sizeof(struct smoothcontrol_state_internal));
```
sizeof(*state)

mlyle reviewed (commented):
Looks good.  I will make another careful pass through this after changes (since this is super-critical).

glowtape committed:
smoothcontrol: Fix integrator timeout.

Tick counter only counts up to 255 and no further (doesn't roll over), only gets zeroed on manual control update. Fixed comparison to make timeout functional.

glowtape committed:
smoothcontrol: Change what sizeof statement refers to.

Double asterisk, because smoothcontrol_state is a typedef on "smoothcontrol_state_internal*".

glowtape commented:
Will be interesting to prove an effect. On these ones attached, which is ActuatorCommand.Roll, with 75% duty cycle on the prediction, grabbed at 2ms from the GCS, you can tell, but only because I'm running an 8th order lowpass at 5hz on the gyros. At normal settings, any sloping goes under in the gyro noise that feeds into ActuatorCommand (at least visually).

(--edit: Altho these were small stick movements.)

![chamfer](https://cloud.githubusercontent.com/assets/4010813/25253296/97217868-2620-11e7-8b35-46e82a3c50eb.png)
